Sources/Client/WebSocket/Provider/URLSessionWebSocketProvider.swift

```diff
@@ -0,0 +1,95 @@
+//
+//  URLSessionWebSocketProvider.swift
+//  StreamChatClient
+//
+//  Created by Alexey Bukhtin on 30/04/2020.
+//  Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+@available(iOS 13, *)
+final class URLSessionWebSocketProvider: NSObject, WebSocketProvider, URLSessionDataDelegate, URLSessionWebSocketDelegate {
+    private var task: URLSessionWebSocketTask?
+    let request: URLRequest
+    var isConnected = false
+    let callbackQueue: DispatchQueue
+    weak var delegate: WebSocketProviderDelegate?
+    
+    init(request: URLRequest, callbackQueue: DispatchQueue) {
+        self.request = request
+        self.callbackQueue = callbackQueue
+    }
+    
+    func connect() {
+        let session = URLSession(configuration: URLSessionConfiguration.default, delegate: self, delegateQueue: nil)
+        task = session.webSocketTask(with: request)
+        doRead()
+        task?.resume()
+    }
+    
+    func disconnect() {
+        isConnected = false
+        task?.cancel(with: .abnormalClosure, reason: nil)
+    }
+    
+    func sendPing() {
+        task?.sendPing { _ in }
+    }
+    
+    private func doRead() {
+        task?.receive { [weak self] result in
+            guard let self = self else {
+                return
+            }
+            
+            switch result {
+            case .success(let message):
+                if case .string(let string) = message {
+                    self.callDelegateInCallbackQueue { $1?.websocketDidReceiveMessage($0, message: string) }
+                }
+            case .failure(let error):
+                self.disconnect(with: WebSocketProviderError(reason: error.localizedDescription,
+                                                             code: (error as NSError).code,
+                                                             providerType: URLSessionWebSocketProvider.self,
+                                                             providerError: error))
+            }
+            
+            self.doRead()
+        }
+    }
+    
+    public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
+        isConnected = true
+        callDelegateInCallbackQueue { $1?.websocketDidConnect($0) }
```

Comment:
🤯

Reply:
I'm sorry. It doesn't look good. It's internal work, but I agree at least I should add comments.
I'm thinking to remove a provider parameter from `WebSocketProviderDelegate` and it will simplify a lot.